clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -488,7 +491,13 @@ private void close(KafkaChannel channel) {
         } catch (IOException e) {
             log.error("Exception closing connection to node {}:", channel.id(), e);
         }
-        this.stagedReceives.remove(channel);
+        Deque<NetworkReceive> deque = this.stagedReceives.remove(channel);
+        if (deque != null) {
+            while (!deque.isEmpty()) {
+                addToCompletedReceives(channel, deque);
+            }
+            closedChannels.put(channel.id(), channel);
```

Comment:
It is probably worth a comment somewhere explaining why we both to keep track of closed (especially in regard to the acks=0 case).